### Решение оптимизационных задач SciPy

In [2]:
from scipy import optimize

In [3]:
# Определим фунцию на которой будем все тестирвоать (функция розенброка)
def f(x):
    return 5 *(1-x[0])**2 + (x[1] - x[0]**2)**2

# Минимум данной функции находиться в продолговатом желобе который не просто найти (значение (1,1))

Примитивный метод: перебор

In [5]:
res = optimize.brute(f, ((-5,5),(-5,5)))
print(res)

[1.00001461 1.00002194]


Если функция негладкая и имеются разрывы, то неплохим методом модет быть **дифференциальная эволюция (генетический алгоритм)**

In [7]:
res = optimize.differential_evolution(f, ((-5,5),(-5,5)))
print(res)

     fun: 2.9582283945787943e-31
 message: 'Optimization terminated successfully.'
    nfev: 3333
     nit: 110
 success: True
       x: array([1., 1.])


Минимум был найден, однако потребывалось не малое количество итераций: 110. Иногда, число итераций может быть очень важным фактором.

Если функция имеет градиент, то можно воспользоваться **градиентными методами**

In [28]:
import numpy as np

# Определим градиент функции розенброка
def g(x):
    return np.array((-2*0.5*(1-x[0]) - 4*x[0]*(x[1] - x[0]**2), 2*(x[1] - x[0]**2)))

Возможно определенный нами градиент неверен, на этот случай есть функция ```check_grad()``` проверяющая соответствие между истинным градиентом функции и найденный нами

In [30]:
print(optimize.check_grad(f, g, [2,2]))

9.000000238418579


Видим, что расхождение не малое, но в пределах нормы. Один из популярных градиентных методов ```bfgs()```

In [31]:
res = optimize.fmin_bfgs(f, [2,2], fprime=g) # fprime - градиент
print(res)

Optimization terminated successfully.
         Current function value: 0.000000
         Iterations: 8
         Function evaluations: 9
         Gradient evaluations: 9
[1.00001245 1.00002597]


Видно, что градиентные методы позволяют существенно быстрей осуществить оптимизацию функций. Поэтому если функция гладкая, то лучше использовать градиентные методы, иначе неградиентные. Воспользуемся методом ```minimize()```

In [33]:
# minimize по умолчанию
res = optimize.minimize(f, [2,2])
print(res)

      fun: 1.4104470358988942e-15
 hess_inv: array([[0.09998587, 0.20001781],
       [0.20001781, 0.89991211]])
      jac: array([-2.70582226e-10, -1.76561210e-10])
  message: 'Optimization terminated successfully.'
     nfev: 24
      nit: 7
     njev: 8
   status: 0
  success: True
        x: array([0.99999998, 0.99999996])


In [34]:
# minimize c параметрами
res = optimize.minimize(f, [2,2], method='BFGS', jac=g)
print(res)

      fun: 7.761940447549734e-10
 hess_inv: array([[0.92044459, 1.83416471],
       [1.83416471, 4.15430675]])
      jac: array([8.18511160e-06, 2.13259657e-06])
  message: 'Optimization terminated successfully.'
     nfev: 9
      nit: 8
     njev: 9
   status: 0
  success: True
        x: array([1.00001245, 1.00002597])


Другой неплохой метод это метод **Nelder Mead**

In [35]:
res = optimize.minimize(f, [2,2], method='Nelder-Mead')
print(res)

 final_simplex: (array([[1.00000795, 0.99997549],
       [0.99997041, 0.99990086],
       [0.99997159, 1.00003024]]), array([1.94877855e-09, 5.97464432e-09, 1.16128206e-08]))
           fun: 1.948778549145414e-09
       message: 'Optimization terminated successfully.'
          nfev: 78
           nit: 42
        status: 0
       success: True
             x: array([1.00000795, 0.99997549])
